xc/xenops_server_xen.ml

```diff
@@ -1796,6 +1796,7 @@ module VM = struct
                    if not(wait_shutdown task vm Suspend 1200.)
                    then raise (Xenopsd_error (Failed_to_shutdown(vm.Vm.id, 1200.)));
                 );
+              post_suspend_callback task;
```

Comment:
I think that this is too late. For live migration, this function has two distinct phases: the live phase (before suspending) and the stop/copy phase (after suspending). We need to make the task uncancellable _just before_ entering the second phase. We can do this at the beginning of the suspend callback function that is the last argument to `Domain.suspend` above, just before `request_shutdown`.

Reply:
Actually, we probably still want to be able to cancel the suspend itself, but not the memory copying that happens after it. So we could disable cancellation just after `wait_shutdown` (so two lines up from here, inside the callback).

Reply:
By the way, this should be manually tested to make sure that the behaviour in the two phases is as expected.

Reply:
Agreed - the task should become uncancellable just before we ask the domain to suspend.

Reply:
Sorry, confusion... In my second comment above I meant to not make the task uncancellable until after `wait_shutdown`, because this may block for up to 20 minutes if the guest has ack'ed but then got stuck. I think that we still want to be able to cancel that bit.

Reply:
I was aware of that you mean. However if we prohibit the cancellation after sending the shutdown request to VM, we have no chance to cancel the task until performing next atomic which will check cancelling at https://github.com/xapi-project/xenopsd/blob/master/lib/xenops_server.ml#L1090.
So what we have to do is before we suspending the VM (in the stop and copy phase):
1. we check cancelling, raise exception if cancelling
2. make the task to not cancellable
The actions above should wrapped in a lock together.

At the same time, we have to update the Task_server.cancel to put the cancellation check into the locked zone.

Reply:
Code updated accordingly.

Reply:
Regarding what you wrote under 'PS': Don't forget that `force_shutdown` only works by cancelling all ongoing tasks, so that won't work until we implement some sort of `Task.force_cancel`

Reply:
So no matter where we put the code to prohibit the cancellation of migration, once the guest hangs after acking the shutdown request, we are impossible to cancel the task immediately until the timeout exceed (which may take 20 mins in the worst case).
Base on this situation, prohibit the cancellation before suspending is a VM is better because starts from this moment if the user cancels migration, the VM will finally get halted.

Reply:
Cancel a task should be a clean operation which will just rollback the actions we have taken before the cancellation, it won't impact the object we intend to operate.
In this case, starts from shutting down the source domain, the migration becomes not able to clean rollback, so we should mark the task not cancellable before requesting shut down the domain.
@jonludlam @robhoes any more comments?